File: app/src/main/java/com/game/tama/command/Command.java
package com.game.tama.command;

import android.util.Log;

import com.game.engine.DisplayAdapter;
import com.game.tama.command.lambda.CommandUpdatable;
import com.game.tama.core.thing.pet.Pet;
import com.game.tama.core.world.World;

import java.io.Serializable;

public abstract class Command implements Serializable
{
    public static final CommandUpdatable noop = (p, w) -> {};

    public CommandUpdatable update = this::start;
    public CommandState state = CommandState.ready;

    protected Pet actor = null;

    protected void start(final Pet pet, final World world)
    {
        actor = pet;
        state = CommandState.doing;
        update = this::doing;
    }

    protected abstract void doing(Pet pet, World world);

    public boolean isReplaceable()
    {
        return state == CommandState.ready;
    }

    public void hardCancel()
    {
        state = CommandState.failed;
        update = noop;
        Log.d(getClass().getCanonicalName(), "canceled");
    }

    public void draw(final DisplayAdapter display) {}

    protected void fail()
    {
        state = CommandState.failed;
        update = noop;
        Log.d(getClass().getCanonicalName(), "failed");
    }

    protected void complete()
    {
        state = CommandState.complete;
        update = noop;
    }
}

File: app/src/main/java/com/game/tama/command/Command.kt
package com.game.tama.command

import android.util.Log
import com.game.android.DisplayAdapter
import com.game.tama.core.World
import com.game.tama.thing.Pet
import java.io.Serializable

/**
 * An abtraction for any single action that a pet can execute
 */
abstract class Command : Serializable
{
    public var update: (pet: Pet, world: World) -> Unit = ::start;
    public var state = CommandState.ready;
    var actor: Pet? = null;

    protected open fun start(pet: Pet, world: World)
    {
        actor = pet;
        state = CommandState.doing;
        update = ::doing;
        
    }

    protected abstract fun doing(pet: Pet, world: World);

    open fun isReplaceable(): Boolean
    {
        return state == CommandState.ready;
    }

    open fun hardCancel()
    {
        state = CommandState.failed;
        update = ::noop;
        Log.d(javaClass.canonicalName, "canceled");
    }

    abstract fun draw(d: DisplayAdapter);

    fun fail()
    {
        state = CommandState.failed;
        update = ::noop
        Log.d(javaClass.canonicalName, "failed");
    }

    fun complete()
    {
        state = CommandState.complete
        update = ::noop
    }

    final fun noop(pet: Pet, world: World){}
}